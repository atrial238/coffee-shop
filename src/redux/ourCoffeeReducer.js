import {getAllCoffeeAPI, getCoffeeByCountryAPI, getCoffeeByNameAPI} from '../API/api';

const setIsError = (bool) => ({type: types.SET_IS_ERROR, body: bool}),
		setIsLoading = (bool) => ({type: types.SET_IS_LOADING, body: bool}),
		setBestCoffee = (data) => ({type: types.GET_COFFEE, body: data}),
		setIsFirstPage = (bool) => ({type: types.SET_IS_FIRST_PAGE, body: bool}),
		setIsLastPage = (bool) => ({type: types.SET_IS_LAST_PAGE, body: bool}),
		setIsSearchAllCoffee = (bool) => ({type: types.SET_IS_SEARCH_ALL_COFFEE, body: bool}),
		setIsSearchByName = (bool) => ({type: types.SET_IS_SEARCH_BY_NAME, body: bool}),
		setIsSearchByCountry = (bool) => ({type: types.SET_IS_SEARCH_BY_COUNTRY, body: bool}),
		setCountry  = (data) => ({type: types.SET_COUNTRY, body: data}),
		setNameCoffee = (data) => ({type: types.SET_NAME_COFFEE, body: data}),
		setPageZero = () => ({type: types.SET_PAGE_ZERO}),
		setIsItemFound = (bool) => ({type: types.SET_IS_ITEM_FOUND, body: bool});

export const setPage = (counter) => ({type: types.SET_PAGE, body: counter});

export const types = {
	GET_COFFEE: 'ourCoffee_reducer/GET_COFFEE',
	SET_IS_LOADING: 'ourCoffee_reducer/SET_IS_LOADING',
	SET_IS_ERROR: 'ourCoffee_reducer/SET_IS_ERROR',
	SET_PAGE: 'ourCoffee_reducer/SET_PAGE',
	SET_IS_FIRST_PAGE: 'ourCoffee_reducer/SET_IS_FIRST_PAGE',
	SET_IS_LAST_PAGE: 'ourCoffee_reducer/SET_IS_LAST_PAGE',
	SET_IS_SEARCH_ALL_COFFEE: 'ourCoffee_reducer/SET_IS_SEARCH_ALL_COFFEE',
	SET_IS_SEARCH_BY_NAME: 'ourCoffee_reducer/SET_IS_SEARCH_BY_NAME',
	SET_IS_SEARCH_BY_COUNTRY: 'ourCoffee_reducer/SET_IS_SEARCH_BY_COUNTRY',
	SET_COUNTRY: 'ourCoffee_reducer/SET_COUNTRY',
	SET_NAME_COFFEE: 'ourCoffee_reducer/SET_NAME_COFFEE',
	SET_PAGE_ZERO: 'ourCoffee_reducer/SET_PAGE_ZERO',
	SET_IS_ITEM_FOUND: 'ourCoffee_reducer/SET_IS_ITEM_FOUND',
}
 

// thunk creator for searching by country
export const searchByCountry = (country) => (dispatch) =>{
	dispatch(setIsSearchByCountry(true));
	dispatch(setIsSearchAllCoffee(false));
	dispatch(setIsSearchByName(false));
	dispatch(setCountry(country));
	dispatch(setPageZero());
}

// thunk creator for searching by name
export const searchByName = (input) => (dispatch) => {
	if(input !== ''){
		dispatch(setIsSearchByCountry(false));
		dispatch(setIsSearchAllCoffee(false));
		dispatch(setIsSearchByName(true));
		dispatch(setPageZero());
		dispatch(setCountry(''));
		dispatch(setNameCoffee(input));
	}
}

// thunk creator for reset filter
export const resetFilter = () => (dispatch) =>{
	dispatch(setIsSearchByCountry(false));
	dispatch(setIsSearchAllCoffee(true));
	dispatch(setIsSearchByName(false));
	dispatch(setPageZero());
	dispatch(setCountry(''));
	dispatch(setNameCoffee(''));
}

//thunk creator make request to a server
export const getCoffee = (page, country, input) => (dispatch, getState) => {
	//destructuring the required variable from the state
	const {ourCoffee: {isSearchAllCoffee, isSearchByName, isSearchByCountry}} = getState();
	
	//method API depends which request must be done
		//1)if get all the coffee without filtering (getAllCoffeeAPI)
		//2)if filter by country (getCoffeeFilteredAPI)
		//3)if filter by name (getCoffeeByNameAPI)
	const methodAPI = (isSearchAllCoffee && getAllCoffeeAPI) 
							|| (isSearchByCountry && getCoffeeByCountryAPI)
							|| (isSearchByName && getCoffeeByNameAPI);
	
	dispatch(setIsFirstPage(false))
	if(page === 0) dispatch(setIsFirstPage(true))
	dispatch(setIsLastPage(false))
	dispatch(setIsError(false))
	dispatch(setIsLoading(true))
	dispatch(setIsItemFound(false))

	methodAPI(page, country, input)
		.then(res => {
			dispatch(setIsLoading(false));
			dispatch(setIsError(false));
			if(res === 'error'){
				dispatch(setIsError(true))
			}else if(isSearchByName && !res.coffeePerPage.length){
				dispatch(setIsItemFound(true))
			}else{
				dispatch(setBestCoffee(res.coffeePerPage))
			}
			if(Math.ceil(res.allItems / 6) === (page + 1)) dispatch(setIsLastPage(true))
		})
}

export const initState = {
	coffeePerPage: [],
	isLoading: false,
	isError: false,
	isFirstPage: false,
	isLastPage: false,
	isSearchAllCoffee: true,
	isSearchByName: false,
	isSearchByCountry: false,
	countrySearch: '',
	nameCoffeeSearch: '',
	page: 0,
	isItemFound: false
}

export const ourCoffeeReducer = (state = initState, action) => {
	switch(action.type){
		case types.GET_COFFEE:
			return {...state, coffeePerPage: action.body};
		case types.SET_IS_LOADING:
			return {...state, isLoading: action.body};
		case types.SET_IS_ERROR:
			return {...state, isError: action.body};
		case types.SET_PAGE:
			return {...state, page: state.page + action.body};
		case types.SET_IS_FIRST_PAGE:
			return {...state, isFirstPage: action.body};
		case types.SET_IS_LAST_PAGE:
			return {...state, isLastPage: action.body};
		case types.SET_IS_SEARCH_ALL_COFFEE:
			return {...state, isSearchAllCoffee: action.body};
		case types.SET_IS_SEARCH_BY_NAME:
			return {...state, isSearchByName: action.body};
		case types.SET_IS_SEARCH_BY_COUNTRY:
			return {...state, isSearchByCountry: action.body};
		case types.SET_COUNTRY:
			return {...state, countrySearch: action.body};
		case types.SET_NAME_COFFEE: 
			return {...state, nameCoffeeSearch: action.body};
		case types.SET_IS_ITEM_FOUND:
			return {...state, isItemFound: action.body};
		case types.SET_PAGE_ZERO:
			return {...state, page: 0};
		default: 
			return state;
	}
}