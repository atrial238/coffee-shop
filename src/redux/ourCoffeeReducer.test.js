import { ourCoffeeReducer, types, initState, searchByCountry, searchByName, resetFilter} from './ourCoffeeReducer';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
 
const mockStore = configureMockStore([thunk]);



describe('ourCoffeeReducer', () => {
	
	test('should return initial state', () => {
		const initialState = ourCoffeeReducer(undefined, {});
		expect(initialState).toMatchSnapshot();
	});

	test('should set coffePerPage array', () => {
		const action =  {type: types.GET_COFFEE, body: [{coffeeone: 'one'}, {coffeeTwo: 'two'}]};
		const state = ourCoffeeReducer(initState, action);
		expect(state).toMatchSnapshot();
	})

	test('should set isLoading true', () => {
		const action =  {type: types.SET_IS_LOADING, body: true};
		const state = ourCoffeeReducer(initState, action);
		expect(state).toMatchSnapshot();
	})

	test('should set isError true', () => {
		const action = {type: types.SET_IS_ERROR, body: true};
		const state = ourCoffeeReducer(initState, action);
		expect(state).toMatchSnapshot();
	})

	test('should decrease on increasing number of page', () => {
		const action = {type: types.SET_PAGE, body: 1};
		const state = ourCoffeeReducer(initState, action);
		expect(state).toMatchSnapshot();
	})

	test('should set first page true', () => {
		const action = {type: types.SET_IS_FIRST_PAGE, body: true};
		const state = ourCoffeeReducer(initState, action);
		expect(state).toMatchSnapshot();
	})

	test('should set last page true', () => {
		const action = {type: types.SET_IS_LAST_PAGE, body: true};
		const state = ourCoffeeReducer(initState, action);
		expect(state).toMatchSnapshot();
	})

	test('should set isSearchAllCoffee true', () => {
		const action = {type: types.SET_IS_SEARCH_ALL_COFFEE, body: true};
		const state = ourCoffeeReducer(initState, action);
		expect(state).toMatchSnapshot();
	})

	test('should set isSearchByName true', () => {
		const action = {type: types.SET_IS_SEARCH_BY_NAME, body: true};
		const state = ourCoffeeReducer(initState, action);
		expect(state).toMatchSnapshot();
	})

	test('should set isSearchByCountry true', () => {
		const action = {type: types.SET_IS_SEARCH_BY_COUNTRY, body: true};
		const state = ourCoffeeReducer(initState, action);
		expect(state).toMatchSnapshot();
	})

	test('should set country for search', () => {
		const action = {type: types.SET_COUNTRY, body: 'country'};
		const state = ourCoffeeReducer(initState, action);
		expect(state).toMatchSnapshot();
	})

	test('should set nameCoffee for search', () => {
		const action = {type: types.SET_NAME_COFFEE, body: 'someName'};
		const state = ourCoffeeReducer(initState, action);
		expect(state).toMatchSnapshot();
	})

	test('should set page zero', () => {
		const action = {type: types.SET_PAGE_ZERO};
		const state = ourCoffeeReducer(initState, action);
		expect(state).toMatchSnapshot();
	})

	test('should set is any coffee found true', () => {
		const action = {type: types.SET_IS_ITEM_FOUND, body: true};
		const state = ourCoffeeReducer(initState, action);
		expect(state).toMatchSnapshot();
	})

})

describe('thunks ourCoffeeReducer', () => {
	let store;

	beforeEach(() => store = mockStore({}));

	it('dispatch action to store for searchByCountry thunk', () => {
		const expectedActions = [
			{type: types.SET_IS_SEARCH_BY_COUNTRY, body: true},
			{type: types.SET_IS_SEARCH_ALL_COFFEE, body: false},
			{type: types.SET_IS_SEARCH_BY_NAME, body: false},
			{type: types.SET_COUNTRY, body: 'country'},
			{type: types.SET_PAGE_ZERO},
		]
	
		store.dispatch(searchByCountry('country'))
		expect(store.getActions()).toEqual(expectedActions);
	});

	it('dispatch action to store for searchByName thunk', () => {

		const expectedActions = [
			{type: types.SET_IS_SEARCH_BY_COUNTRY, body: false},
			{type: types.SET_IS_SEARCH_ALL_COFFEE, body: false},
			{type: types.SET_IS_SEARCH_BY_NAME, body: true},
			{type: types.SET_PAGE_ZERO},
			{type: types.SET_COUNTRY, body: ''},
			{type: types.SET_NAME_COFFEE, body: 'someName'},
		]

		store.dispatch(searchByName('someName'));
		expect(store.getActions()).toEqual(expectedActions);
	});

	it('dispatch action to store for resetFilter thunk', () => {

		const expectedActions = [
			{type: types.SET_IS_SEARCH_BY_COUNTRY, body: false},
			{type: types.SET_IS_SEARCH_ALL_COFFEE, body: true},
			{type: types.SET_IS_SEARCH_BY_NAME, body: false},
			{type: types.SET_PAGE_ZERO},
			{type: types.SET_COUNTRY, body: ''},
			{type: types.SET_NAME_COFFEE, body: ''},
		]

		store.dispatch(resetFilter('someName'));
		expect(store.getActions()).toEqual(expectedActions);
	});
 })
 