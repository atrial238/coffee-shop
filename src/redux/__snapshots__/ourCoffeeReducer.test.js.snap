// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`homeReducer shoud return initial state 1`] = `
Object {
  "bestCoffe": Array [],
  "isError": false,
  "isLoading": false,
}
`;

exports[`homeReducer shoud set bestCoffee array 1`] = `
Object {
  "bestCoffe": Array [
    Object {
      "coffeeone": "one",
    },
    Object {
      "coffeeTwo": "two",
    },
  ],
  "isError": false,
  "isLoading": false,
}
`;

exports[`homeReducer shoud set isError true 1`] = `
Object {
  "bestCoffe": Array [],
  "isError": true,
  "isLoading": false,
}
`;

exports[`homeReducer shoud set isLoading true 1`] = `
Object {
  "bestCoffe": Array [],
  "isError": false,
  "isLoading": true,
}
`;

exports[`ourCoffeeReducer should decrease on increasing number of page 1`] = `
Object {
  "coffeePerPage": Array [],
  "countrySearch": "",
  "isError": false,
  "isFirstPage": false,
  "isItemFound": false,
  "isLastPage": false,
  "isLoading": false,
  "isSearchAllCoffee": true,
  "isSearchByCountry": false,
  "isSearchByName": false,
  "nameCoffeeSearch": "",
  "page": 1,
}
`;

exports[`ourCoffeeReducer should return initial state 1`] = `
Object {
  "coffeePerPage": Array [],
  "countrySearch": "",
  "isError": false,
  "isFirstPage": false,
  "isItemFound": false,
  "isLastPage": false,
  "isLoading": false,
  "isSearchAllCoffee": true,
  "isSearchByCountry": false,
  "isSearchByName": false,
  "nameCoffeeSearch": "",
  "page": 0,
}
`;

exports[`ourCoffeeReducer should set coffePerPage array 1`] = `
Object {
  "coffeePerPage": Array [
    Object {
      "coffeeone": "one",
    },
    Object {
      "coffeeTwo": "two",
    },
  ],
  "countrySearch": "",
  "isError": false,
  "isFirstPage": false,
  "isItemFound": false,
  "isLastPage": false,
  "isLoading": false,
  "isSearchAllCoffee": true,
  "isSearchByCountry": false,
  "isSearchByName": false,
  "nameCoffeeSearch": "",
  "page": 0,
}
`;

exports[`ourCoffeeReducer should set country for search 1`] = `
Object {
  "coffeePerPage": Array [],
  "countrySearch": "country",
  "isError": false,
  "isFirstPage": false,
  "isItemFound": false,
  "isLastPage": false,
  "isLoading": false,
  "isSearchAllCoffee": true,
  "isSearchByCountry": false,
  "isSearchByName": false,
  "nameCoffeeSearch": "",
  "page": 0,
}
`;

exports[`ourCoffeeReducer should set first page true 1`] = `
Object {
  "coffeePerPage": Array [],
  "countrySearch": "",
  "isError": false,
  "isFirstPage": true,
  "isItemFound": false,
  "isLastPage": false,
  "isLoading": false,
  "isSearchAllCoffee": true,
  "isSearchByCountry": false,
  "isSearchByName": false,
  "nameCoffeeSearch": "",
  "page": 0,
}
`;

exports[`ourCoffeeReducer should set is any coffee found true 1`] = `
Object {
  "coffeePerPage": Array [],
  "countrySearch": "",
  "isError": false,
  "isFirstPage": false,
  "isItemFound": true,
  "isLastPage": false,
  "isLoading": false,
  "isSearchAllCoffee": true,
  "isSearchByCountry": false,
  "isSearchByName": false,
  "nameCoffeeSearch": "",
  "page": 0,
}
`;

exports[`ourCoffeeReducer should set isError true 1`] = `
Object {
  "coffeePerPage": Array [],
  "countrySearch": "",
  "isError": true,
  "isFirstPage": false,
  "isItemFound": false,
  "isLastPage": false,
  "isLoading": false,
  "isSearchAllCoffee": true,
  "isSearchByCountry": false,
  "isSearchByName": false,
  "nameCoffeeSearch": "",
  "page": 0,
}
`;

exports[`ourCoffeeReducer should set isLoading true 1`] = `
Object {
  "coffeePerPage": Array [],
  "countrySearch": "",
  "isError": false,
  "isFirstPage": false,
  "isItemFound": false,
  "isLastPage": false,
  "isLoading": true,
  "isSearchAllCoffee": true,
  "isSearchByCountry": false,
  "isSearchByName": false,
  "nameCoffeeSearch": "",
  "page": 0,
}
`;

exports[`ourCoffeeReducer should set isSearchAllCoffee true 1`] = `
Object {
  "coffeePerPage": Array [],
  "countrySearch": "",
  "isError": false,
  "isFirstPage": false,
  "isItemFound": false,
  "isLastPage": false,
  "isLoading": false,
  "isSearchAllCoffee": true,
  "isSearchByCountry": false,
  "isSearchByName": false,
  "nameCoffeeSearch": "",
  "page": 0,
}
`;

exports[`ourCoffeeReducer should set isSearchByCountry true 1`] = `
Object {
  "coffeePerPage": Array [],
  "countrySearch": "",
  "isError": false,
  "isFirstPage": false,
  "isItemFound": false,
  "isLastPage": false,
  "isLoading": false,
  "isSearchAllCoffee": true,
  "isSearchByCountry": true,
  "isSearchByName": false,
  "nameCoffeeSearch": "",
  "page": 0,
}
`;

exports[`ourCoffeeReducer should set isSearchByName true 1`] = `
Object {
  "coffeePerPage": Array [],
  "countrySearch": "",
  "isError": false,
  "isFirstPage": false,
  "isItemFound": false,
  "isLastPage": false,
  "isLoading": false,
  "isSearchAllCoffee": true,
  "isSearchByCountry": false,
  "isSearchByName": true,
  "nameCoffeeSearch": "",
  "page": 0,
}
`;

exports[`ourCoffeeReducer should set last page true 1`] = `
Object {
  "coffeePerPage": Array [],
  "countrySearch": "",
  "isError": false,
  "isFirstPage": false,
  "isItemFound": false,
  "isLastPage": true,
  "isLoading": false,
  "isSearchAllCoffee": true,
  "isSearchByCountry": false,
  "isSearchByName": false,
  "nameCoffeeSearch": "",
  "page": 0,
}
`;

exports[`ourCoffeeReducer should set nameCoffee for search 1`] = `
Object {
  "coffeePerPage": Array [],
  "countrySearch": "",
  "isError": false,
  "isFirstPage": false,
  "isItemFound": false,
  "isLastPage": false,
  "isLoading": false,
  "isSearchAllCoffee": true,
  "isSearchByCountry": false,
  "isSearchByName": false,
  "nameCoffeeSearch": "someName",
  "page": 0,
}
`;

exports[`ourCoffeeReducer should set page zero 1`] = `
Object {
  "coffeePerPage": Array [],
  "countrySearch": "",
  "isError": false,
  "isFirstPage": false,
  "isItemFound": false,
  "isLastPage": false,
  "isLoading": false,
  "isSearchAllCoffee": true,
  "isSearchByCountry": false,
  "isSearchByName": false,
  "nameCoffeeSearch": "",
  "page": 0,
}
`;
